src/Grpc.AspNetCore.Server/GrpcEndpointRouteBuilderExtensions.cs

```diff
@@ -20,60 +20,94 @@
-                var baseType = service.BaseType;
-                while (baseType?.BaseType?.BaseType != null)
-                {
-                    baseType = baseType.BaseType;
-                }
-
-                // We need to call Foo.BindService from the declaring type.
-                var declaringType = baseType?.DeclaringType;
+                options.BindAction(serviceBinder, null);
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidOperationException($"Error binding gRPC service '{typeof(TService).Name}'.", ex);
+            }
 
-                // The method we want to call is public static void BindService(ServiceBinderBase, BaseType)
-                var bindService = declaringType?.GetMethod("BindService", new[] { typeof(ServiceBinderBase), baseType });
+            return new CompositeEndpointConventionBuilder(serviceBinder.EndpointConventionBuilders);
+        }
 
-                if (bindService == null)
-                {
-                    throw new InvalidOperationException($"Cannot locate BindService(ServiceBinderBase, ServiceBase) method for the current service type: {service.FullName}. The type must be an implementation of a gRPC service.");
-                }
+        private static void ReflectionBind<TService>(ServiceBinderBase binder, TService _)
+        {
+            var service = typeof(TService);
 
-                var callHandlerFactory = builder.ServiceProvider.GetRequiredService<ServerCallHandlerFactory<TService>>();
+            // TService is an implementation of the gRPC service. It ultimately derives from Foo.TServiceBase base class.
+            // We need to access the static BindService method on Foo which implicitly derives from Object.
+            var baseType = service.BaseType;
+            while (baseType?.BaseType?.BaseType != null)
+            {
+                baseType = baseType.BaseType;
+            }
 
-                var serviceBinder = new GrpcServiceBinder<TService>(builder, callHandlerFactory);
+            // We need to call Foo.BindService from the declaring type.
+            var declaringType = baseType?.DeclaringType;
 
-                // Invoke
-                bindService.Invoke(null, new object[] { serviceBinder, null });
+            // The method we want to call is public static void BindService(ServiceBinderBase, BaseType)
+            var bindService = declaringType?.GetMethod("BindService", new[] { typeof(ServiceBinderBase), baseType });
 
-                return new CompositeEndpointConventionBuilder(serviceBinder.EndpointConventionBuilders);
-            }
-            catch (Exception ex)
+            if (bindService == null)
             {
-                throw new InvalidOperationException($"Unable to map gRPC service '{service.Name}'.", ex);
+                throw new InvalidOperationException($"Cannot locate BindService(ServiceBinderBase, ServiceBase) method for the current service type: {service.FullName}.");
```

Comment:
Why remove "The type must be an implementation of a gRPC service."? If ReflectionBind is being used, TService does need to implement a generated gRPC service type, right?

Reply:
Because of different code gen output, what we support as a gRPC service is different from what your tool output as a gRPC service. This error message is more accurate to what the problem is.